man/systemd.service.xml

```diff
@@ -897,6 +897,28 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RestartMode=</varname></term>
+
+        <listitem>
+          <para>Takes a string value that specifies how a service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed/inactive state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service restarts without
+              going through a failed/inactive state.
+              <varname>ExecStopPost=</varname> is invoked.
+              <varname>OnFailure=</varname> is skipped.</para></listitem>
```

Comment:
OnSuccess= is skipped too, also deserves mention.

Reply:
Actually `OnSuccess=` is called when the service goes from the `auto-restart` state to the `dead` state

Reply:
hmm?

Since a1d315730ffddf283d4bb9d73878fbcd97a4d244 we definitely go into low-level state     SERVICE_DEAD_BEFORE_AUTO_RESTART (which maps to high-level UNIT_DEAD) before entering auto-restart. This should trigger the OnSuccess= dependency whenever we terminate a service successfully and then restart.

Reply:
Yes, this is what I see. Here are the states we go through for a service exiting cleanly and with `Restart=always`
- `RestartMode=normal` : `running -> stop -> stop-post -> dead-before-auto-restart -> auto-restart -> dead -> running`
- `RestartMode=quick` : `running -> stop -> stop-post -> auto-restart -> dead -> running`

For a failing service:
- `RestartMode=normal` : `running -> stop-post -> failed-before-auto-restart -> auto-restart -> dead -> running`
- `RestartMode=quick` : `running -> stop-post -> auto-restart -> dead -> running`

Reply:
So let's mention here that `OnSuccess=` is skipped. Yes it is still called due to #27722 but that's a bug that shouldn't be taken into account here.

Reply:
And that's already fixed by #28215.